### PR TITLE
FED-466: Fix console logs that require formatting

### PR DESCRIPTION
**Pull request by kealjones-wk**
## Motivation
  
These logs suck and are no help: 
Screen Shot 2022-10-14 at 2 18 39 PM

## Changes
  
- Add a dart "formatter" (based on an [old version of node js's console format](https://github.com/nodejs/node-v0.x-archive/blob/master/lib/console.js).... the [newer one is scary])(https://github.com/nodejs/node/blob/main/lib/internal/util/inspect.js#L2173-L2287)
  - Result: 
Screen Shot 2022-10-14 at 2 19 05 PM

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author:
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/react_testing_library/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/react_testing_library/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/util/console_log_utils.dart [hunk 1]
```diff
@@ -152,3 +154,51 @@ class ConsoleConfig {
...
+      if (x == '%%') return '%';
+      if (i >= len) return x;
+      switch (x) {
+        case '%s':
+          return args[i++].toString();
+        case '%d':
+          return num.tryParse(args[i++].toString()).toString();
+        case '%j':
+          try {
+            return callMethod(getProperty(window,'JSON'), 'stringify', [args[i++]]).toString();
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_utils.dart:186]: Could we use static JS interop here instead?


```dart
@JS('JSON.stringify')
external String _jsonStringify(dynamic object);
```

```suggestion
            return _jsonStringify(args[i++]);
```

File: lib/src/util/console_log_utils.dart [hunk 2]
```diff
@@ -152,3 +154,51 @@ class ConsoleConfig {
...
 }
+
+final formatRegExp = RegExp('%[sdj%]');
+String format(dynamic f, List<dynamic> arguments) {
+  if (f is! String) {
+    final objects = [];
+    for (var i = 0; i < arguments.length; i++) {
+      objects.add(arguments[i]);
+    }
+    return objects.join(' ');
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_utils.dart:165]:
```suggestion
    return [f, ...arguments].join(' ');
```

File: lib/src/util/console_log_utils.dart [hunk 3]
```diff
@@ -152,3 +154,51 @@ class ConsoleConfig {
...
+  }
+
+  var str = '';
+  if (f is String) {
+    if (!f.contains('%')) return f;
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_utils.dart:170]: This discards `arguments` when `f` doesn't have anything that should be interpolated, as opposed to concatenating them; could we add a test case for that?

File: lib/src/util/console_log_utils.dart [hunk 4]
```diff
@@ -152,3 +154,51 @@ class ConsoleConfig {
...
+
+    var i = 0;
+    final args = arguments;
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_utils.dart:173]: #nit this variable is unnecessary; why not just use `arguments`?

File: lib/src/util/console_log_utils.dart [hunk 5]
```diff
@@ -152,3 +154,51 @@ class ConsoleConfig {
...
+    final len = args.length;
+    str += f.replaceAllMapped(formatRegExp, (m) {
+      final x = m[0].toString();
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_utils.dart:176]: #nit unnecessary `toString`

File: lib/src/util/console_log_utils.dart [hunk 6]
```diff
@@ -152,3 +154,51 @@ class ConsoleConfig {
...
+      switch (x) {
+        case '%s':
+          return args[i++].toString();
+        case '%d':
+          return num.tryParse(args[i++].toString()).toString();
+        case '%j':
+          try {
+            return callMethod(getProperty(window,'JSON'), 'stringify', [args[i++]]).toString();
+          } catch (_) {
+            return '[Circular]';
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_utils.dart:188]: #nit in this case could we just return `x`?

File: test/unit/console_log_utils_test.dart [hunk 7]
```diff
@@ -130,6 +131,19 @@ void main() {
...
       });
 
+      test('prints logs that use formatter syntax', () {
+        // This also tests functionally that the print call occurs in the right zone,
+        // which need to happen for them to be forwarded in the terminal in a test environment.
+        // If it wasn't in the right zone, we wouldn't be able to record it, and neither would
+        // the test package.
+        final printCalls = recordPrintCalls(() {
+          printConsoleLogs(() {
+            callMethod(getProperty(window, 'console'), 'log', ['%s World Number %d! %j', 'Hello', 5, jsify({'doWeComeInPeace':false}), 'additional']);
```
Comment (greglittlefield-wf) [inline test/unit/console_log_utils_test.dart:141]: Can we also add a case without `'additional'` so we have test coverage for that codepath?

File: lib/src/util/console_log_utils.dart [hunk 8]
```diff
@@ -152,3 +154,51 @@ class ConsoleConfig {
...
+          break;
+        default:
+          return x;
+      }
+    });
+
+    if (i < len) {
+      for (var x = args[i]; i < len; x = args[i++]) {
+        str += ' $x';
+      }
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_utils.dart:199]: The control flow on this is kind of hard to follow; what about something like:
```suggestion
      str += ' ' + args.skip(i).join(' ');
```

File: lib/src/util/console_log_utils.dart [hunk 9]
```diff
@@ -152,3 +154,51 @@ class ConsoleConfig {
   /// Captures calls to `console.log`, `console.warn`, and `console.error`.
   static const ConsoleConfig all = ConsoleConfig._('all');
 }
+
+final formatRegExp = RegExp('%[sdj%]');
+String format(dynamic f, List<dynamic> arguments) {
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_utils.dart:159]: This could use a doc comment

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Some comments, but this looks awesome!!

File: lib/src/util/console_log_utils.dart (hunk 1 above)
Comment (kealjones-wk) [reply lib/src/util/console_log_utils.dart:186]: ugh but that would mean i have to like make this a library, give it a name, and add package js, ugh im exhausted already

File: lib/src/util/console_log_utils.dart (hunk 2 above)
Comment (kealjones-wk) [reply lib/src/util/console_log_utils.dart:165]: I didn't pop the previous arguments off mate

File: lib/src/util/console_log_utils.dart (hunk 3 above)
Comment (kealjones-wk) [reply lib/src/util/console_log_utils.dart:170]: mmmmmmm good point, good point

File: lib/src/util/console_log_utils.dart (hunk 4 above)
Comment (kealjones-wk) [reply lib/src/util/console_log_utils.dart:173]: i legit copy pasted and converted the node js impl i have no idea.

File: lib/src/util/console_log_utils.dart (hunk 2 above)
Comment (greglittlefield-wf) [reply lib/src/util/console_log_utils.dart:165]: What do you mean?

File: lib/src/util/console_log_utils.dart (hunk 4 above)
Comment (greglittlefield-wf) [reply lib/src/util/console_log_utils.dart:173]: Can you please attribute that code then? 😛

File: lib/src/util/console_log_utils.dart (hunk 6 above)
Comment (kealjones-wk) [reply lib/src/util/console_log_utils.dart:188]: how would you know it was circular json then?

File: test/unit/console_log_utils_test.dart (hunk 7 above)
Comment (kealjones-wk) [reply test/unit/console_log_utils_test.dart:141]: you mean like when there isn't enough args?

File: lib/src/util/console_log_utils.dart (hunk 2 above)
Comment (kealjones-wk) [reply lib/src/util/console_log_utils.dart:165]: i dont know

File: test/unit/console_log_utils_test.dart (hunk 7 above)
Comment (greglittlefield-wf) [reply test/unit/console_log_utils_test.dart:141]: That's a good case too, but I was thinking when there are exactly the number of args to match the format placeholders, like this case without `'additional'`

Comment (kealjones-wk) [reply test/unit/console_log_utils_test.dart:141]: wow

File: lib/src/util/console_log_formatter.dart [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+library console_log_formatter;
+
+import 'package:js/js.dart';
+
+@JS('JSON.stringify')
+external String _jsonStringify(dynamic object);
+
+final _formatRegExp = RegExp('%[sdj%]');
+
+/// A doc comment
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_formatter.dart:35]: You want me to close this PR? 😛

File: test/unit/console_log_utils_test.dart [hunk 11]
```diff
@@ -131,17 +131,50 @@ void main() {
...
-            callMethod(getProperty(window, 'console'), 'log', ['%s World Number %d! %j', 'Hello', 5, jsify({'doWeComeInPeace':false}), 'additional']);
+      group('prints logs that use formatter syntax', () {
+        test('with exact amount of args', () {
+          // This also tests functionally that the print call occurs in the right zone,
+          // which need to happen for them to be forwarded in the terminal in a test environment.
+          // If it wasn't in the right zone, we wouldn't be able to record it, and neither would
+          // the test package.
+          final printCalls = recordPrintCalls(() {
+            printConsoleLogs(() {
+              callMethod(getProperty(window, 'console'), 'log', ['%s World Number %d! %j', 'Hello']);
```
Comment (greglittlefield-wf) [inline test/unit/console_log_utils_test.dart:142]: Looks like this is missing the jsified argument, the number, and the additional arg

File: lib/src/util/console_log_formatter.dart [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+      if (x == '%%') return '%';
+      if (i >= len) return x;
+      switch (x) {
+        case '%s':
+          return arguments[i++].toString();
+        case '%d':
+          return num.tryParse(arguments[i++].toString()).toString();
+        case '%j':
+          try {
+            return _jsonStringify([arguments[i++]]);
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_formatter.dart:56]: Since this isn't callMethod, it shouldn't be wrapped in a list.
```suggestion
            return _jsonStringify(arguments[i++]);
```

File: test/unit/console_log_utils_test.dart [hunk 13]
```diff
@@ -223,8 +267,7 @@ void main() {
     if (runtimeSupportsPropTypeWarnings()) {
       group('captures errors correctly', () {
         test('when mounting', () {
-          final logs = recordConsoleLogs(() => rtl.render(Sample({'shouldAlwaysBeFalse': true}) as ReactElement),
-              configuration: ConsoleConfig.error);
+          final logs = recordConsoleLogs(() => rtl.render(Sample({'shouldAlwaysBeFalse': true}) as ReactElement), configuration: ConsoleConfig.error);
```
Comment (greglittlefield-wf) [inline test/unit/console_log_utils_test.dart:270]: Looks like these got formatted differently by mistake

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]

File: lib/src/util/console_log_formatter.dart (hunk 10 above)
Comment (kealjones-wk) [reply lib/src/util/console_log_formatter.dart:35]: 🖕 get out of here friday greg its wednesday.

Comment (greglittlefield-wf) [reply lib/src/util/console_log_formatter.dart:35]: This is worse than no doc comment lol

File: test/unit/console_log_utils_test.dart (hunk 13 above)
Comment (kealjones-wk) [reply test/unit/console_log_utils_test.dart:270]: they bug me... so i re-formatted them in the next commit.

File: lib/src/util/console_log_formatter.dart [hunk 14]
```diff
@@ -53,7 +55,9 @@ String format(dynamic f, List<dynamic> arguments) {
           return num.tryParse(arguments[i++].toString()).toString();
         case '%j':
           try {
-            return _jsonStringify([arguments[i++]]);
+            final argToStringify = arguments[i++];
+            debugger();
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_formatter.dart]: Looks like this and the import were committed by mistake.

File: lib/src/util/console_log_utils.dart [hunk 15]
```diff
@@ -91,26 +99,37 @@ void Function() startSpyingOnConsoleLogs({
...
   for (final config in logTypeToCapture) {
-    consoleRefs[config] = context['console'][config] as JsFunction;
-    context['console'][config] = JsFunction.withThis((self, [message, arg1, arg2, arg3, arg4, arg5]) {
-      // NOTE: Using console.log or print within this function will cause an infinite
-      // loop when the logType is set to `log`.
-      boundOnLog(format(message?.toString(), [arg1, arg2, arg3, arg4, arg5]));
-      consoleRefs[config].apply([message, arg1, arg2, arg3, arg4, arg5], thisArg: self);
-    });
+    consolePropertyDescriptors[config] = _getOwnPropertyDescriptor(_console, config);
+    consoleRefs[config] = context['console'][config];
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_utils.dart:112]: Could we restore the cast that was previously here along with the typing on `consoleRefs`?

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Just a couple more things

File: lib/src/util/console_log_utils.dart [hunk 16]
```diff
@@ -152,3 +184,30 @@ class ConsoleConfig {
   /// Captures calls to `console.log`, `console.warn`, and `console.error`.
   static const ConsoleConfig all = ConsoleConfig._('all');
 }
+
+dynamic _jsFunctionApply(dynamic jsFunction, List<dynamic> args) {
+  return callMethod(jsFunction, 'apply', [null, jsify(args)]);
```
Comment (greglittlefield-wf) [inline lib/src/util/console_log_utils.dart:189]: #nit Is there a reason we don't pass in a `this` arg (from `allowInteropCaptureThis`) instead of `null` here, like we were doing before?

It looks like it works without it in Chrome, but it seems like it'd be nice to have to be safe.

**Review by greglittlefield-wf — APPROVED** [review]
+10

File: lib/src/util/console_log_utils.dart (hunk 16 above)
Comment (kealjones-wk) [reply lib/src/util/console_log_utils.dart:189]: That was what i did first but when i tried that it had issues, i don't recall exactly what it was but then i tried with null instead and it was passing so i went with it. lol

Comment (kealjones-wk) [reply lib/src/util/console_log_utils.dart]: and of course i try it again and it works...

**Review by greglittlefield-wf — APPROVED** [review]
+10 assuming CI is all green

**Comment by greglittlefield-wf** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM
